internals-js/src/specs/sourceSpec.ts

```diff
@@ -0,0 +1,146 @@
+  InputFieldDefinition,
+  ListType,
+} from "../definitions";
+import { registerKnownFeature } from '../knownCoreFeatures';
+import { createDirectiveSpecification } from '../directiveAndTypeSpecification';
+
+export const sourceIdentity = 'https://specs.apollo.dev/source';
+
+export class SourceSpecDefinition extends FeatureDefinition {
+  constructor(version: FeatureVersion, minimumFederationVersion?: FeatureVersion) {
+    super(new FeatureUrl(sourceIdentity, 'source', version), minimumFederationVersion);
+
+    this.registerDirective(createDirectiveSpecification({
+      name: 'sourceAPI',
+      locations: [DirectiveLocation.SCHEMA],
+      repeatable: true,
+      // We "compose" these `@source{API,Type,Field}` directives using the
+      // `@join__directive` mechanism, so they do not need to be composed in the
+      // way passing `composes: true` here implies.
+      composes: false,
+    }));
+
+    this.registerDirective(createDirectiveSpecification({
+      name: 'sourceType',
+      locations: [DirectiveLocation.OBJECT, DirectiveLocation.INTERFACE],
+      repeatable: true,
+      composes: false,
+    }));
+
+    this.registerDirective(createDirectiveSpecification({
+      name: 'sourceField',
+      locations: [DirectiveLocation.FIELD_DEFINITION],
+      repeatable: true,
+      composes: false,
+    }));
+  }
+
+  addElementsToSchema(schema: Schema): GraphQLError[] {
+    const sourceAPI = this.addDirective(schema, 'sourceAPI').addLocations(DirectiveLocation.SCHEMA);
+    sourceAPI.repeatable = true;
+
+    sourceAPI.addArgument('name', new NonNullType(schema.stringType()));
+
+    const HTTPHeaderMapping = schema.addType(new InputObjectType('HTTPHeaderMapping'));
+    HTTPHeaderMapping.addField(new InputFieldDefinition('name')).type =
+      new NonNullType(schema.stringType());
+    HTTPHeaderMapping.addField(new InputFieldDefinition('as')).type =
+      schema.stringType();
+    HTTPHeaderMapping.addField(new InputFieldDefinition('value')).type =
+      schema.stringType();
+
+    const HTTPSourceAPI = schema.addType(new InputObjectType('HTTPSourceAPI'));
+    HTTPSourceAPI.addField(new InputFieldDefinition('baseURL')).type =
+      new NonNullType(schema.stringType());
+    HTTPSourceAPI.addField(new InputFieldDefinition('headers')).type =
+      new ListType(new NonNullType(HTTPHeaderMapping));
+    sourceAPI.addArgument('http', HTTPSourceAPI);
+
+    const sourceType = this.addDirective(schema, 'sourceType').addLocations(
+      DirectiveLocation.OBJECT,
+      DirectiveLocation.INTERFACE,
+      // TODO Allow @sourceType on unions, similar to interfaces?
+      // DirectiveLocation.UNION,
+    );
+    sourceType.repeatable = true;
+    sourceType.addArgument('api', new NonNullType(schema.stringType()));
+
+    const URLPathTemplate = this.addScalarType(schema, 'URLPathTemplate');
+    const JSONSelection = this.addScalarType(schema, 'JSONSelection');
+    const JSON = this.addScalarType(schema, 'JSON');
```

Comment:
What would happen if a JSON definition already existed in the schema? Would it be safer to namespace the scalars to avoid conflicting definintions?

Reply:
I think I used to have bigger plans for this `JSON` than the one place it's used now: the `typeMap` field of the `KeyTypeMap` input object type. Given it's only used there, I'll adjust the scalar typename to match that use case.

Reply:
https://github.com/apollographql/federation/pull/2894/commits/d6f7e8a5815ca65fc7150aaafcc38c8cf5baa810